### ISSUE TITLE
Configure kube-reserved-cgroup on nodes so to enable --kube-reserved on Rancher Kubernetes

### ISSUE THREAD

**Issue by windswordsgd** [rancher/rancher#17177]
**What kind of request is this (question/bug/enhancement/feature request):**
Questions

**Steps to reproduce (least amount of steps as possible):**
I have set the kube-reserved flag in kubelet. However, it won't work if a kube-served CG group is not specified. I checked the Rancher Kubernetes and it does not create a new CG group automatically after updating cluster.config.  Refer to https://github.com/rancher/rancher/issues/17154

Kubelet is containerized on the nodes.  Are there any way or the instruction to create cgroup on each node for kube-reserved cgroup?

Refer to
https://github.com/kubernetes/community/blob/master/contributors/design-proposals/node/node-allocatable.md#recommended-cgroups-setup

Specifying kube-reserved value in this flag is invalid if --kube-reserved-cgroup flag is not specified.
--kube-reserved-cgroup=
This flag helps kubelet identify the control group managing all kube components like Kubelet & container runtime that fall under the KubeReserved reservation.
Example: /kube.slice. Note that absolute paths are required and systemd naming scheme isn't supported.

**Other details that may be helpful:**

Environment information

Rancher version (rancher/rancher/rancher/server image tag or shown bottom left in the UI):
Installation option (single install/HA):
Single
Cluster information

Cluster type (Hosted/Infrastructure Provider/Custom/Imported): Rancher Kubernetes on Azure VM ( Not AKS)
Machine type (cloud/VM/metal) and specifications (CPU/memory): VM 2C4G
Kubernetes version (use kubectl version):
Client Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.0", GitCommit:"91e7b4fd31fcd3d5f436da26c980becec37ceefe", GitTreeState:"clean", BuildDate:"2018-06-27T20:17:28Z", GoVersion:"go1.10.2", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.5", GitCommit:"32ac1c9073b132b8ba18aa830f46b77dcceb0723", GitTreeState:"clean", BuildDate:"2018-06-21T11:34:22Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}

Docker (use docker info):
Containers: 26
Running: 10
Paused: 0
Stopped: 16
Images: 6
Server Version: 17.03.2-ce
Storage Driver: aufs
Root Dir: /var/lib/docker/aufs
Backing Filesystem: extfs
Dirs: 106
Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
Volume: local
Network: bridge host macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 4ab9917febca54791c5f071a9d1f404867857fcc
runc version: 54296cf40ad8143b62dbcaa1d90e520a2136ddfe
init version: 949e6fa
Security Options:
apparmor
seccomp
Profile: default
Kernel Version: 4.15.0-1021-azure
Operating System: Ubuntu 16.04.5 LTS
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 3.853 GiB
Name: davidk8smaster1
ID: 2EYI:ACJ5:I6EV:6476:SLJI:UWIT:ID3H:SLMZ:X4T3:5TZ3:M3QN:QB7T
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
provider=azure
Experimental: false
Insecure Registries:
127.0.0.0/8
Live Restore Enabled: false

**Comment by windswordsgd** [comment]
Is there any update? Any advice is much appreciated

**Comment by loganhz** [comment]
How did you configure `kube-reserved-cgroup` ? 
You can go to edit cluster page and click edit as yaml to add the param. 
https://rancher.com/docs/rke/v0.1.x/en/config-options/services/services-extras/#extra-args

Then please go to the node to find out if `kube-reserved-cgroup` has been added when rancher starts kubelet

**Comment by dabeck** [comment]
@loganhz The question is which cgroup should be configured?

**Comment by garyhuang123** [comment]
i'm having the same issue what's the value of kube-reserved-cgroup in Rancher? Since the kubelet is now running in container rather than on host.

**Comment by loganhz** [comment]
![image](https://user-images.githubusercontent.com/21168270/53586127-7be7e500-3bc2-11e9-8b5f-b390dcfdd749.png)
![image](https://user-images.githubusercontent.com/21168270/53586180-95892c80-3bc2-11e9-9d15-f00e12e6ebfa.png)

It works for me. @dabeck @garyhuang123 @windswordsgd 

Please make sure you have `enforce-node-allocatable`, `kube-reserved-cgroup` and `kube-reserved` configured correctly.

RKE already mounts the cgroup path. 

But yes, you have to create your cgroup by yourself as k8s docs said. 
For example,

`mkdir -p /sys/fs/cgroup/hugetlb/mykube`
`mkdir -p /sys/fs/cgroup/cpu/mykube`
`mkdir -p /sys/fs/cgroup/cpuacct/mykube`
`mkdir -p /sys/fs/cgroup/cpuset/mykube`
`mkdir -p /sys/fs/cgroup/memory/mykube`
`mkdir -p /sys/fs/cgroup/systemd/mykube`

https://github.com/kubernetes/community/blob/master/contributors/design-proposals/node/node-allocatable.md#recommended-cgroups-setup
https://github.com/kubernetes/kubernetes/blob/v1.13.1/pkg/kubelet/cm/cgroup_manager_linux.go#L259

**Comment by willmao** [comment]
@loganhz 
 
according to the k8s doc below

``
Kube-Reserved (proposed) - Compute resources reserved for Kubernetes components such as the docker daemon, kubelet, kube proxy, etc.
``

the cgroup for kube-reserved is not used only for the kubelet process, but also other kubernetes components like docker daemon, kube proxy and nginx-proxy set up by rke.

Does rke use the `kube-reserved-cgroup` parameter set in kubelet section to setup kube-proxy or nginx-proxy ?

**Comment by Symbianx** [comment]
Would it work if we had a DaemonSet execute the mkdir commands and then just hang until restarted? Or do the cgroups have to be created before docker starting?

**Comment by KaelBaldwin** [comment]
Running on ubuntu virtual machines for nodes, I haven't been able to find any way to set things up as recommended here: https://github.com/kubernetes/community/blob/master/contributors/design-proposals/node/node-allocatable.md#recommended-cgroups-setup

@loganhz your instructions perhaps work if I was using the default cgroup driver, but I believe it's recommended to only use either systemd or the cgroup driver and not both due to some instability issues. Though even with your instructions, I'm not sure how all of the other kubernetes components could be set to the same cgroup as @willmao points out.

The closest I've been able to get is to tell docker to use systemd as the driver and kubectl to use systemd as well. Once I did that then kubernetes is running all of it's pods in the kubepods.slice, all the other docker and system processes are running in system.slice, and there is user.slice as well for user processes.

Based on the above, I set the kube-reserved-cgroup as /system.slice and the system-reserved-cgroup as /user.slice which has at least fixed the stability issues I've been having where the kubelet has been locking up on my cluster.

I would really like to be able to set things up the recommended way, but I've tried every kubectl extra_arg I could find and haven't been able to get it to work. Has anyone been able to get this to fully work as recommended with rancher?

Thank you.

**Comment by dmelha** [comment]
Is there any update on this? 

I tried using the steps suggested in [@loganhz's ](https://github.com/rancher/rancher/issues/17177#issuecomment-468368299)** reply but failed (on Ubuntu 1804 - I first tried using docker+cgroupfs  and kubelet using cgroupdriver=cgroupfs; then I tried docker+snative.cgroupdriver=systemd and kubelet using cgroupdriver=systemd  but failed too.

Is there documentation anywhere to get this to work (on Ubuntu1804)?

**Comment by dmelha** [comment]
to answer my own question:
tldr; `pids` control group is/was missing from loganhz's example given above for use on ubuntu1804 with cgroupdriver defaults.

for ubuntu1804 with kubelet AND docker using cgroupdriver **cgroupfs** this works:
`
mkdir -p /sys/fs/cgroup/pids/mykube
mkdir -p /sys/fs/cgroup/hugetlb/mykube
mkdir -p /sys/fs/cgroup/cpuset/mykube
mkdir -p /sys/fs/cgroup/cpu/mykube
mkdir -p /sys/fs/cgroup/memory/mykube
`
alternatively:
`
apt-get install cgroup-tools
cgcreate -g pids,hugetlb,cpuset,cpu,memory:mykube
`

**Comment by bobbydheep** [comment]
Hi All, 
I am trying to achieve this kube-reserved configs in our rancher setup as well. After I edit the cluster with the reserved configd for kubelet, I see the reduced allocatable resources for the node. But after that the cluster is unstable and everything resumes only until I revoke the changes. Could you please advice /anyone if you have any better experiences so far. 

FATA[0073] Failed to bring up Worker Plane: [Failed to verify healthcheck: Failed to check http://localhost:10256/healthz for service [kube-proxy]

Plus :  As we are editing the cluster config here and not on the nodes directly ( like on direct kubernetes servers,  kubelete service running as a system service and not as container ), how does this really affect any active cluster, do you need to delete all nodes and re-register? (I tried this also, but no luck).

Thanks
Dheep

**Comment by bobbydheep** [comment]
Hi Guys,  

I just figured out my problems, the cgroup schema/names "xxx.slice" were not being discovered by the kubelet. I corrected this and now we have kube-reserved and system reserved resources set along with eviction thresholds for available memory and available nodefs. The kubepods.slices are auto-created by kubelet container. 
        eviction-hard: 'memory.available<500Mi,nodefs.available<10%'
        kube-reserved: 'cpu=1,memory=1Gi,ephemeral-storage=1Gi'
        system-reserved: 'cpu=1,memory=1Gi,ephemeral-storage=1Gi'
        kube-reserved-cgroup: /kubepods.slice
        system-reserved-cgroup: /system.slice
        enforce-node-allocatable: 'pods,kube-reserved,system-reserved'

Now the nodes and cluster are updated fine and we have 
[Allocatable] = [Node Capacity] - [Kube-Reserved] - [System-Reserved] - [Hard-Eviction-Threshold]. 
If you dont plan to fine control/limit on the system resources or add reserves to enforceNodeAllocatable, then no need to specify cgroups. Not recommended unless you have a very metered usage pattern in your environment. So if you follow the kubernetes link:  [https://kubernetes.io/docs/tasks/administer-cluster/reserve-compute-resources/#general-guidelines] //_Be extra careful while enforcing system-reserved reservation since it can lead to critical system services being CPU starved, OOM killed, or unable to fork on the node. The recommendation is to enforce system-reserved only if a user has profiled their nodes exhaustively to come up with precise estimates and is confident in their ability to recover if any process in that group is oom_killed. To begin with enforce Allocatable on pods._  


Hope it helps somebody !!

**Comment by AntonSmolkov** [comment]
>  However, it won't work if a kube-served CG group is not specified

Actually, both `--system-reserved` and `--kube-reserved` work without cgroups definition: 
![image](https://user-images.githubusercontent.com/5318028/90414757-ddeb2180-e0b8-11ea-8794-abff256177f5.png)
You have to set `--(system|kube)-reserved-cgroup` only if you'd like to **LIMIT** resource consumption of system/kube processes. 


Copy/Pasting the [documentation ](https://kubernetes.io/docs/tasks/administer-cluster/reserve-compute-resources/#general-guidelines)here:

> - To begin with enforce Allocatable on pods.
>- Once adequate monitoring and alerting is in place to track kube system daemons, attempt to enforce kube-reserved based on usage heuristics.
>- If absolutely necessary, enforce system-reserved over time.




I personally fine with just fencing a little resources from pods.

**Comment by Turb0Fly** [comment]
> > However, it won't work if a kube-served CG group is not specified
> 
> Actually, both `--system-reserved` and `--kube-reserved` work without cgroups definition:
> ![image](https://user-images.githubusercontent.com/5318028/90414757-ddeb2180-e0b8-11ea-8794-abff256177f5.png)
> You have to set `--(system|kube)-reserved-cgroup` only if you'd like to **LIMIT** resource consumption of system/kube processes.
> 
> Copy/Pasting the [documentation ](https://kubernetes.io/docs/tasks/administer-cluster/reserve-compute-resources/#general-guidelines)here:
> 
> > * To begin with enforce Allocatable on pods.
> > * Once adequate monitoring and alerting is in place to track kube system daemons, attempt to enforce kube-reserved based on usage heuristics.
> > * If absolutely necessary, enforce system-reserved over time.
> 
> I personally fine with just fencing a little resources from pods.

By implementing this, you do not limit system and kube processes ? You just take away those resources from allocatable pods on each node ?
I have an issue where nodes were getting unstable due to a high memory consumption. I added the kubelet flag for hard eviction and it seemed to help stabilize the worker nodes, however I still see warning messages with kubelet not having enough memory.

**Comment by weisdd** [comment]
@AntonSmolkov Thanks for the example, though I'd say it's better not to leave out the default `eviction-hard` settings, because the values get replaced, not merged. Here's an improved version of the setting (everything else is omitted):

```yaml
rancher_kubernetes_engine_config:
  services:
    kubelet:
      extra_args:
         eviction-hard: "memory.available<500Mi,imagefs.available<15%,nodefs.available<10%,nodefs.inodesFree<5%"
```
